lite2/client.go

```diff
@@ -0,0 +1,419 @@
+// bisection algorithm from the specification is used for finding the minimal
+// "trust path".
+//
+// trustLevel - fraction of the old validator set (in terms of voting power),
+// which must sign the new header in order for us to trust it. NOTE this only
+// applies to non-adjusted headers. For adjusted headers, sequential
+// verification is used.
+func SkippingVerification(trustLevel cmn.Fraction) Option {
+	if err := ValidateTrustLevel(trustLevel); err != nil {
+		panic(err)
+	}
+	return func(c *Client) {
+		c.verificationMode = skipping
+		c.trustLevel = trustLevel
+	}
+}
+
+// AlternativeSources option can be used to supply alternative providers, which
+// will be used for cross-checking the primary provider.
+func AlternativeSources(providers []provider.Provider) Option {
+	return func(c *Client) {
+		c.alternatives = providers
+	}
+}
+
+// Client represents a light client, connected to a single chain, which gets
+// headers from a primary provider, verifies them either sequentially or by
+// skipping some and stores them in a trusted store (usually, a local FS).
+//
+// Default verification: SkippingVerification(DefaultTrustLevel)
+type Client struct {
+	chainID          string
+	trustingPeriod   time.Duration // see TrustOptions.Period
+	verificationMode mode
+	trustLevel       cmn.Fraction
+
+	// Primary provider of new headers.
+	primary provider.Provider
+
+	// Alternative providers for checking the primary for misbehavior by
+	// comparing data.
+	alternatives []provider.Provider
+
+	// Where trusted headers are stored.
+	trustedStore    store.Store
+	trustedHeader   *types.SignedHeader // H
+	trustedNextVals *types.ValidatorSet // H+1
+
+	logger log.Logger
+}
+
+// NewClient returns a new light client. It returns an error if it fails to
+// obtain the header & vals from the primary or they are invalid (e.g. trust
+// hash does not match with the one from the header).
+//
+// See all Option(s) for the additional configuration.
+func NewClient(
+	chainID string,
+	trustOptions TrustOptions,
+	primary provider.Provider,
+	trustedStore store.Store,
+	options ...Option) (*Client, error) {
+
+	c := &Client{
+		chainID:          chainID,
+		trustingPeriod:   trustOptions.Period,
+		verificationMode: skipping,
+		trustLevel:       DefaultTrustLevel,
+		primary:          primary,
+		trustedStore:     trustedStore,
+		logger:           log.NewNopLogger(),
+	}
+
+	for _, o := range options {
+		o(c)
+	}
+
+	if err := c.initializeWithTrustOptions(trustOptions); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Client) initializeWithTrustOptions(options TrustOptions) error {
+	// 1) Fetch and verify the header.
+	h, err := c.primary.SignedHeader(options.Height)
+	if err != nil {
+		return err
+	}
+
+	// NOTE: Verify func will check if it's expired or not.
+	if err := h.ValidateBasic(c.chainID); err != nil {
+		return errors.Wrap(err, "ValidateBasic failed")
+	}
+
```

Comment:
Should we fetch also c.primary.ValidatorSet(options.Height) and make sure 2/3rds signed?
Is there a way to have a function similar to verifyNewHeaderAndVals() that does not make the "h1" checks there? i.e. given a signed header and the valset for that header do all checks wether it's for the first header used to initialize the client, or for subsequent verifications.

Reply:
> Should we fetch also c.primary.ValidatorSet(options.Height) and make sure 2/3rds signed?

what for? thought the whole point of trusting a header with HASH XYZ is that you trust it. Plus we will need to make 1 more request.

> Is there a way to have a function similar to verifyNewHeaderAndVals() that does not make the "h1" checks there?

see verifyNewHeaderAndVals in verifier.go

Reply:
yes, but that takes h1 as parameter too.

Reply:
>> Should we fetch also c.primary.ValidatorSet(options.Height) and make sure 2/3rds signed?
>
>what for? thought the whole point of trusting a header with HASH XYZ is that you trust it. Plus we will need to make 1 more request.

We already do some checks on this "trusted" header. I was wondering why not the >2/3rds signed one. And what would happen if this would fail, vs. what happens when the basic checks fail. We definitely have to "blindly" trust the validator set that signed this header. 
I am also not entirely sure if getting the header from the primary provider is the way to initialize the client (vs. passing the header to the client init function and leave out of scope of this PR how it is obtained). I must admit I don't know the precise mechanism of "social consensus".

Reply:
> And what would happen if this would fail, vs. what happens when the basic checks fail

light client exits with error?

On the second thought, we should probably design light client for max security. at least initially and only after maybe relax conditions. I will add the check.

Reply:
https://github.com/tendermint/tendermint/pull/4209/commits/e757cff5a707718e2ebd6d9a7384738822f7edfd